### PR TITLE
don't use panics.

### PR DESCRIPTION
**Pull request by n3wscott**
Too many of these errors:

```
    --- FAIL: TestBrokerDirect/Assert/[Alpha/MUST]RabbitMQ_source_the_recorder_received_all_sent_events_within_the_time (240.00s)
panic: Fail in goroutine after TestBrokerDirect/Setup/install_recorder has completed [recovered]
	panic: Fail in goroutine after TestBrokerDirect/Setup/install_recorder has completed
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -295,7 +295,7 @@ func Install(message string) feature.StepFn {
 	}
 	return func(ctx context.Context, t *testing.T) {
 		if _, err := manifest.InstallLocalYaml(ctx, cfg); err != nil {
-			t.Fatal(err)
+			t.Error(err)
```
Comment (slinkydeveloper) [inline README.md:298]: This has to panic :smile:

File: pkg/eventshub/event_info_store.go [hunk 2]
```diff
@@ -158,7 +158,8 @@ func (ei *Store) Find(matchers ...EventInfoMatcher) ([]eventshub.EventInfo, even
 func (ei *Store) AssertAtLeast(min int, matchers ...EventInfoMatcher) []eventshub.EventInfo {
 	events, err := ei.waitAtLeastNMatch(allOf(matchers...), min)
 	if err != nil {
-		ei.t.Fatalf("Timeout waiting for at least %d matches.\nError: %+v", min, errors.WithStack(err))
+		ei.t.Errorf("Timeout waiting for at least %d matches.\nError: %+v", min, errors.WithStack(err))
```
Comment (slinkydeveloper) [inline pkg/eventshub/event_info_store.go:161]: This has to panic

**Review by slinkydeveloper — COMMENTED** [review]
All the panics (which are not panics, but t.fatal) here are correct, they have to panic. If you remove the panic, it will just make harder for us to find out why they panic :smile:

File: README.md (hunk 1 above)
Comment (n3wscott) [reply README.md:298]: panic here kills the entire test run

Comment (slinkydeveloper) [reply README.md:298]: It must kill the feature run

**Comment by n3wscott** [comment]
Using fatals in the step mean the tests end early and the test run does not attempt the later asserts.

**Comment by slinkydeveloper** [comment]
> Using fatals in the step mean the tests end early and the test run does not attempt the later asserts.

That is the expected behaviour... If you want to change this behaviour, you need to do it at the level of abstraction discussed in #106, not the single fatal calls here

**Comment by n3wscott** [comment]
> That is the expected behaviour... If you want to change this behaviour, you need to do it at the level of abstraction discussed in #106, not the single fatal calls here

To change how this works is changing how `go test` works, I would rather make an example of not using fatals for non fatal conditions in the tests. `4 != 5` is not a fatal condition, but it is a failure; don't panic the rest of the test because a condition failed.

**Comment by slinkydeveloper** [comment]
Run the "should fail" tests in this PR https://github.com/knative-sandbox/reconciler-test/pull/142

**Comment by n3wscott** [comment]
> Run the "should fail" tests in this PR #142

This violates the runtime contract of `go test` is what I am trying to explain. Test authors should use fatal or panics when it truly is fatal and the test should not keep going. #142 changes the meaning of fatal for testing in go, and that is a mistake.

**Comment by slinkydeveloper** [comment]
Sorry wrong button :)

File: README.md (hunk 1 above)
Comment (n3wscott) [reply README.md:298]: ok this one maybe is ok

**Pull request closed without merging**